template/base/miner.py

```diff
@@ -0,0 +1,222 @@
+            priority_fn=self.priority,
+        )
+        bt.logging.info(f"Axon created: {self.axon}")
+
+        # Instantiate runners
+        self.should_exit: bool = False
+        self.is_running: bool = False
+        self.thread: threading.Thread = None
+        self.lock = asyncio.Lock()
+
+    def run(self):
+        """
+        Initiates and manages the main loop for the miner on the Bittensor network.
+
+        This function performs the following primary tasks:
+        1. Check for registration on the Bittensor network.
+        2. Attaches the miner's forward, blacklist, and priority functions to its axon.
+        3. Starts the miner's axon, making it active on the network.
+        4. Regularly updates the metagraph with the latest network state.
+        5. Optionally sets weights on the network, defining how much trust to assign to other nodes.
+        6. Handles graceful shutdown on keyboard interrupts and logs unforeseen errors.
+
+        The miner continues its operations until `should_exit` is set to True or an external interruption occurs.
+        During each epoch of its operation, the miner waits for new blocks on the Bittensor network, updates its
+        knowledge of the network (metagraph), and sets its weights. This process ensures the miner remains active
+        and up-to-date with the network's latest state.
+
+        Note:
+            - The function leverages the global configurations set during the initialization of the miner.
+            - The miner's axon serves as its interface to the Bittensor network, handling incoming and outgoing requests.
+
+        Raises:
+            KeyboardInterrupt: If the miner is stopped by a manual interruption.
+            Exception: For unforeseen errors during the miner's operation, which are logged for diagnosis.
+        """
+
+        # Serve passes the axon information to the network + netuid we are hosting on.
+        # This will auto-update if the axon port of external ip have changed.
+        bt.logging.info(
+            f"Serving miner axon {self.axon} on network: {self.config.subtensor.chain_endpoint} with netuid: {self.config.netuid}"
+        )
+        self.axon.serve(netuid=self.config.netuid, subtensor=self.subtensor)
+
+        # Start  starts the miner's axon, making it active on the network.
+        bt.logging.info(f"Starting axon server on port: {self.config.axon.port}")
+        self.axon.start()
+
+        # --- Run until should_exit = True.
+        bt.logging.info(f"Miner starting at block: {self.block}")
+
+        # This loop maintains the miner's operations until intentionally stopped.
+        bt.logging.info(f"Starting miner main loop")
+        try:
+            while not self.should_exit:
+                while (
+                    self.block - self.metagraph.last_update[self.uid]
+                    < self.config.neuron.epoch_length
+                ):
+                    # --- Wait for next bloc.
```

Comment:
Is this comment necessary and correct? If we want to wait for a block it should be more like 12 seconds.